contracts/external/cw-payroll-factory/src/contract.rs

```diff
@@ -0,0 +1,183 @@
+    TMP_CONTRACT_INFO.save(deps.storage, &(recipient, instantiator))?;
+
+    Ok(Response::default()
+        .add_attribute("action", "instantiate_cw_payroll")
+        .add_submessage(msg))
+}
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn query(deps: Deps, _env: Env, msg: QueryMsg) -> StdResult<Binary> {
+    match msg {
+        QueryMsg::ListVestingContracts { start_after, limit } => {
+            let limit = limit.unwrap_or(DEFAULT_LIMIT).min(MAX_LIMIT) as usize;
+            let start = start_after.as_deref().map(Bound::exclusive);
+
+            let res: Vec<VestingContract> = vesting_contracts()
+                .range(deps.storage, start, None, Order::Ascending)
+                .take(limit)
+                .map(|vc| Ok::<VestingContract, ContractError>(vc?.1))
+                .flatten()
+                .collect();
+
+            Ok(to_binary(&res)?)
+        }
+        QueryMsg::ListVestingContractsByInstantiator {
+            instantiator,
+            start_after,
+            limit,
+        } => {
+            let limit = limit.unwrap_or(DEFAULT_LIMIT).min(MAX_LIMIT) as usize;
+            let start = start_after.map(Bound::<String>::exclusive);
+
+            // Validate owner address
+            deps.api.addr_validate(&instantiator)?;
+
+            let res: Vec<VestingContract> = vesting_contracts()
+                .idx
+                .owner
+                .prefix(instantiator)
+                .range(deps.storage, start, None, Order::Ascending)
+                .take(limit)
+                .map(|vc| Ok::<VestingContract, ContractError>(vc?.1))
+                .flatten()
+                .collect();
+
+            Ok(to_binary(&res)?)
+        }
+        QueryMsg::ListVestingContractsByRecipient {
+            recipient,
+            start_after,
+            limit,
+        } => {
+            let limit = limit.unwrap_or(DEFAULT_LIMIT).min(MAX_LIMIT) as usize;
+            let start = start_after.map(Bound::<String>::exclusive);
+
+            // Validate recipient address
+            deps.api.addr_validate(&recipient)?;
+
+            let res: Vec<VestingContract> = vesting_contracts()
+                .idx
+                .recipient
+                .prefix(recipient)
+                .range(deps.storage, start, None, Order::Ascending)
+                .take(limit)
+                .map(|vc| Ok::<VestingContract, ContractError>(vc?.1))
+                .flatten()
+                .collect();
+
+            Ok(to_binary(&res)?)
+        }
+    }
+}
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn reply(deps: DepsMut, _env: Env, msg: Reply) -> Result<Response, ContractError> {
+    match msg.id {
+        INSTANTIATE_CONTRACT_REPLY_ID => {
+            let res = parse_reply_instantiate_data(msg)?;
+            let contract_addr = deps.api.addr_validate(&res.contract_address)?;
+
+            let (recipient, instantiator) = TMP_CONTRACT_INFO.load(deps.storage)?;
```

Comment:
is there any case in which TMP_CONTRACT_INFO would not correspond to the contract which has just been instantiated?

Reply:
We're passing the `recipient` and `instantiator` through to the submessage here. Not 100% sure if this is the best pattern.

Alternately, we could query the newly created contract... which may be better?

Reply:
I don't think we would be at risk for TMP_CONTRACT_INFO holding incorrect information as in the case of multiple instantiation requests, submessage order will still be preserved, and in the case of an instantiation failure, TMP_CONTRACT_INFO still needs to be set to the correct info before we handle reply success here. querying the created contract would be clearer, but not sure it's entirely necessary

Reply:
@0xekez, any preference or thoughts here?

Reply:
generally speaking, i think it's best to query from the contract in reply > storing state in tmp keys.

while unlikely to happen, this current design is vulnerable to rentrancy where if the vesting contract being instantiated called back into the factory (for some unknown or malicious reason), the tmp variable would get overridden.